app/master/cluster_master.py

```diff
@@ -279,7 +282,7 @@ def handle_result_reported_from_slave(self, slave_url, build_id, subjob_id, payl
         build = self._all_builds_by_id[int(build_id)]
         slave = self._all_slaves_by_url[slave_url]
         # If the build has been canceled, don't work on the next subjob.
-        if not build.is_finished:  # WIP(joey): This check should be internal to the Build object.
+        if not build.is_finished or build.has_error:  # WIP(joey): This check should be internal to the Build object.
```

Comment:
@tjlee0909, yeah your comment sounds plausible. If that's the case I'm not opposed to leaving the has_error check in, even though it is not obvious why it is there. (I think my note about making it internal to the build object is still valid.)

Actually I suspect that `if` check is not even necessary. I bet we could just delete the `if not build.is_finished` line entirely. The point that TJ brought up about slaves not necessarily being torn down/reassigned is a good one and I bet there are some cases where this `if not build.is_finished` check prevents that.

This is definitely tied in with the larger problems with build.cancel that I'm working on, so if you prefer to hand wave it for now I'm fine with that.